File: backend/src/main/java/com/medicoai/service/AppointmentService.java
package com.medicoai.service;

import com.medicoai.dto.request.AppointmentRequest;
import com.medicoai.dto.request.RescheduleRequest;
import com.medicoai.dto.response.AppointmentResponse;
import com.medicoai.enums.AppointmentStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Contract for all appointment business operations.
 */
public interface AppointmentService {

    /** Book a new appointment. patientUserId is the logged-in user's ID. */
    AppointmentResponse book(Long patientUserId, AppointmentRequest request);

    /** Get appointment by ID. */
    AppointmentResponse getById(Long id);

    /** Paginated list with optional status + keyword filter (admin view). */
    Page<AppointmentResponse> getAll(AppointmentStatus status, String search, Pageable pageable);

    /** All appointments for a patient (paginated). */
    Page<AppointmentResponse> getByPatient(Long patientId, Pageable pageable);

    /** All appointments for a doctor (paginated). */
    Page<AppointmentResponse> getByDoctor(Long doctorId, Pageable pageable);

    /** Update notes/reason on a PENDING appointment. */
    AppointmentResponse update(Long id, AppointmentRequest request);

    /** Cancel an appointment with an optional reason. */
    AppointmentResponse cancel(Long id, String reason);

    /** Reschedule to a new date/time. */
    AppointmentResponse reschedule(Long id, RescheduleRequest request);

    /** Doctor confirms a PENDING appointment. */
    AppointmentResponse confirm(Long id);

    /** Doctor marks appointment as COMPLETED. */
    AppointmentResponse complete(Long id);

    // Dashboard counters
    long countTotal();
    long countToday();
    long countByStatus(AppointmentStatus status);
}

File: frontend/src/services/appointmentService.js
import api from './api'

/**
 * appointmentService – all appointment API calls.
 * Matches AppointmentController endpoints exactly.
 */
const appointmentService = {

  /**
   * GET /appointments?status=&search=&page=&size=&sortBy=&sortDir=
   * Returns Page<AppointmentResponse>  (ADMIN / DOCTOR)
   */
  getAll: (params) => api.get('/appointments', { params }),

  /** GET /appointments/{id} */
  getById: (id) => api.get(`/appointments/${id}`),

  /**
   * GET /appointments/patient/{patientId}?page=&size=
   * Returns Page<AppointmentResponse>
   */
  getByPatient: (patientId, params) =>
    api.get(`/appointments/patient/${patientId}`, { params }),

  /**
   * GET /appointments/doctor/{doctorId}?page=&size=
   * Returns Page<AppointmentResponse>
   */
  getByDoctor: (doctorId, params) =>
    api.get(`/appointments/doctor/${doctorId}`, { params }),

  /**
   * POST /appointments
   * Body: { doctorId, patientId?, appointmentDate, appointmentTime,
   *         type, reason, notes }
   */
  book: (data) => api.post('/appointments', data),

  /**
   * PUT /appointments/{id}
   * Body: { reason?, notes?, type? }
   */
  update: (id, data) => api.put(`/appointments/${id}`, data),

  /**
   * PUT /appointments/{id}/cancel
   * Body: { reason? }
   */
  cancel: (id, reason) =>
    api.put(`/appointments/${id}/cancel`, reason ? { reason } : {}),

  /**
   * PUT /appointments/{id}/reschedule
   * Body: { newDate, newTime }
   */
  reschedule: (id, newDate, newTime) =>
    api.put(`/appointments/${id}/reschedule`, { newDate, newTime }),

  /** PUT /appointments/{id}/confirm  (DOCTOR / ADMIN) */
  confirm: (id) => api.put(`/appointments/${id}/confirm`),

  /** PUT /appointments/{id}/complete  (DOCTOR / ADMIN) */
  complete: (id) => api.put(`/appointments/${id}/complete`),
}

export default appointmentService
